### PR TITLE
[GEP-19] Adapt monitoring configuration for shoot control plane components

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area monitoring
/kind enhancement

**What this PR does / why we need it**:
As a follow-up of https://github.com/gardener/gardener/pull/9695 and similar to https://github.com/gardener/gardener/pull/9737, this PR adapts the monitoring configuration for the shoot control plane components running inside the seed cluster.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/9065

**Special notes for your reviewer**:
/cc @ScheererJ 

**Release note**:

```feature developer
NONE
```

### CODE REVIEW

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 1]
```diff
@@ -276,6 +278,76 @@ func (c *clusterAutoscaler) Deploy(ctx context.Context) error {
...
 
+	if _, err := controllerutils.GetAndCreateOrMergePatch(ctx, c.client, prometheusRule, func() error {
+		metav1.SetMetaDataLabel(&prometheusRule.ObjectMeta, "prometheus", shoot.Label)
+		prometheusRule.Spec = monitoringv1.PrometheusRuleSpec{
+			Groups: []monitoringv1.RuleGroup{{
+				Name: "cluster-autoscaler.rules",
+				Rules: []monitoringv1.Rule{{
+					Alert: "ClusterAutoscalerDown",
+					Expr:  intstr.FromString(`absent(up{job="cluster-autoscaler"} == 1)`),
+					For:   ptr.To(monitoringv1.Duration("7m")),
```
Comment (ScheererJ) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: Interesting choice of timeframe, but it is probably scientifically chosen ;-) . Hence, it is fine with me.
I prefer to use number systems based on 2, 8, 10, or 16, but maybe I should diversify.

File: pkg/component/nodemanagement/machinecontrollermanager/machine_controller_manager.go [hunk 2]
```diff
@@ -285,6 +289,115 @@ func (m *machineControllerManager) Deploy(ctx context.Context) error {
...
 
+	if _, err := controllerutils.GetAndCreateOrMergePatch(ctx, m.client, prometheusRule, func() error {
+		metav1.SetMetaDataLabel(&prometheusRule.ObjectMeta, "prometheus", shoot.Label)
+		prometheusRule.Spec = monitoringv1.PrometheusRuleSpec{
+			Groups: []monitoringv1.RuleGroup{{
+				Name: "machine-controller-manager.rules",
+				Rules: []monitoringv1.Rule{{
+					Alert: "MachineControllerManagerDown",
+					Expr:  intstr.FromString(`absent(up{job="machine-controller-manager"} == 1)`),
+					For:   ptr.To(monitoringv1.Duration("15m")),
```
Comment (ScheererJ) [inline pkg/component/nodemanagement/machinecontrollermanager/machine_controller_manager.go:300]: Why is this alert coming so much later than for `cluster-autoscaler` (`7m` vs. `15m`)?

File: pkg/component/gardener/resourcemanager/resource_manager.go [hunk 3]
```diff
@@ -1123,29 +1123,26 @@ func (r *resourceManager) emptyPodDisruptionBudget() *policyv1.PodDisruptionBudg
 	}
 }
 
+func (r *resourceManager) getPrometheusLabel() string {
+	if r.values.TargetDiffersFromSourceCluster {
+		return shoot.Label
+	}
+	return seed.Label
+}
```
Comment (ScheererJ) [inline pkg/component/gardener/resourcemanager/resource_manager.go:1131]: This could be externalized to a helper function is this or something very similar is used in multiple components. However, it is also "just" an `if` statement and two `return`s. Therefore, I am a bit undecided if we should introduce a helper function or not.
WDYT?

File: pkg/component/observability/monitoring/kubestatemetrics/resources.go [hunk 4]
```diff
@@ -484,6 +552,161 @@ func (k *kubeStateMetrics) reconcileScrapeConfigGarden(scrapeConfig *monitoringv
...
+func (k *kubeStateMetrics) emptyScrapeConfigShoot() *monitoringv1alpha1.ScrapeConfig {
+	return &monitoringv1alpha1.ScrapeConfig{ObjectMeta: monitoringutils.ConfigObjectMeta("kube-state-metrics", k.namespace, shoot.Label)}
+}
+
+func (k *kubeStateMetrics) reconcileScrapeConfigShoot(scrapeConfig *monitoringv1alpha1.ScrapeConfig) {
+	scrapeConfig.Labels = monitoringutils.Labels(shoot.Label)
+	scrapeConfig.Spec = k.standardScrapeConfigSpec(
+		"kube_daemonset_metadata_generation",
+		"kube_daemonset_status_current_number_scheduled",
+		"kube_daemonset_status_desired_number_scheduled",
```
Comment (ScheererJ) [inline pkg/component/observability/monitoring/kubestatemetrics/resources.go:564]: Any particular reason why we specify the allowed metrics of the shoot cluster directly inline, but for the cache prometheus we use a variable?

File: pkg/component/etcd/etcd/etcd.go [hunk 5]
```diff
@@ -570,101 +576,333 @@ func (e *etcd) Deploy(ctx context.Context) error {
 			return err
 		}
 	}
-	// etcd deployed for garden cluster
+
+	serviceMonitorJobNameEtcd, serviceMonitorJobNameBackupRestore := "kube-etcd3-"+e.values.Role, "kube-etcd3-backup-restore-"+e.values.Role
 	if e.values.NamePrefix != "" {
```
Comment (ScheererJ) [inline pkg/component/etcd/etcd/etcd.go]: Why did you remove the comment? Is it no longer valid?

**Review by ScheererJ — COMMENTED** [review]
Thanks for the very well structured final addition to GEP-19. Looks very good already.

File: pkg/component/nodemanagement/machinecontrollermanager/machine_controller_manager.go (hunk 2 above)
Comment (rfranzke) [reply pkg/component/nodemanagement/machinecontrollermanager/machine_controller_manager.go:300]: 😄 I don't know. Initially, it was `5m`, see https://github.com/gardener/gardener/commit/e7a1abb0901aa65603f7c1048e454f96b0b9a932.
With https://github.com/gardener/gardener/pull/872, it got changed to `15m`. It is not clear why CA was not adapted as well. I will change it to `15m` as well.

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go (hunk 1 above)
Comment (rfranzke) [reply pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: Hehe, nice comment. Ref https://github.com/gardener/gardener/pull/9848#discussion_r1617014397

File: pkg/component/gardener/resourcemanager/resource_manager.go (hunk 3 above)
Comment (rfranzke) [reply pkg/component/gardener/resourcemanager/resource_manager.go:1131]: I don't think it makes much sense for now since it (a) depends on the values of the component (GRM uses `TargetDiffersFromSourceCluster`), others might use something else, and (b) it might not always be the `seed.Label` but could also be the `aggregate.Label`.

File: pkg/component/observability/monitoring/kubestatemetrics/resources.go (hunk 4 above)
Comment (rfranzke) [reply pkg/component/observability/monitoring/kubestatemetrics/resources.go:564]: For some reasons that I as translator cannot explain, we explicitly specify the list of allowed metrics for the "cache" Prometheus case here: https://github.com/gardener/gardener/blob/c87c89dfb94738584f40aa7d5f823f2273d2f1d4/pkg/component/observability/monitoring/kubestatemetrics/resources.go#L227

We don't do this for the "shoot" Prometheus case. 🤷🏻
